File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        stack = []
        max_water = 0
        for i in range(len(height)):
            while stack and height[i] > height[stack[-1]]:
                mid = stack.pop()
                if not stack:
                    break
                left = stack[-1]
                right = i
                width = right-left-1
                bd = min(height[left],height[right])-height[mid]
                max_water += width*bd
            stack.append(i)
        return max_water

File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int water = 0;
        int left = 0;
        int n = height.size();
        int right = n-1;

        int leftMax = 0;
        int rightMax = 0;

        while(left <= right)
        {
            if(height[left] <= height[right])
            {
                if(height[left] >= leftMax)
                {
                    leftMax = height[left];
                }
                else
                {
                    water += leftMax-height[left];
                }
                left++;
            }
            else
            {
                if(height[right] >= rightMax)
                {
                    rightMax = height[right];
                }
                else
                {
                    water += rightMax-height[right];
                }
                right--;
            }
        }
        return water;
    }
};